server/db/migrate.go

```diff
@@ -0,0 +1,127 @@
+	MigrationsDirectionDown MigrationDirection = "down"
+
+	migrationsTableName = "db_migrations_calls"
+)
+
+// morphLogWriter is a target to pass to the logger instance of morph.
+// For now, everything is just logged at a debug level. If we need to log
+// errors/warnings from the library also, that needs to be seen later.
+type morphLogWriter struct {
+	log mlog.LoggerIFace
+}
+
+func (l *morphLogWriter) Write(in []byte) (int, error) {
+	l.log.Debug(string(in))
+	return len(in), nil
+}
+
+func (s *Store) initMorph(dryRun bool, timeoutSecs int) (*morph.Morph, error) {
+	assetsList, err := assets.ReadDir(path.Join("migrations", s.driverName))
+	if err != nil {
+		return nil, err
+	}
+
+	assetNamesForDriver := make([]string, len(assetsList))
+	for i, entry := range assetsList {
+		assetNamesForDriver[i] = entry.Name()
+	}
+
+	src, err := mbindata.WithInstance(&mbindata.AssetSource{
+		Names: assetNamesForDriver,
+		AssetFunc: func(name string) ([]byte, error) {
+			return assets.ReadFile(path.Join("migrations", s.driverName, name))
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate source assets: %w", err)
+	}
+
+	var driver drivers.Driver
+	switch s.driverName {
+	case model.DatabaseDriverMysql:
+		// MySQL requires the multiStatements flag to be on for migrations to
+		// work so we need to open a dedicated connection.
+
+		dsn, dsnErr := resetReadTimeout(*s.settings.DataSource)
+		if dsnErr != nil {
+			return nil, fmt.Errorf("failed to reset read timeout: %w", dsnErr)
+		}
+		dsn, dsnErr = appendMultipleStatementsFlag(dsn)
+		if err != nil {
+			return nil, fmt.Errorf("failed to append multiple statements flag: %w", dsnErr)
+		}
+
+		db, connErr := s.setupConnForMigrations(dsn)
+		if connErr != nil {
+			return nil, fmt.Errorf("failed to setup db connection: %w", connErr)
+		}
+		defer db.Close()
+
+		driver, err = ms.WithInstance(db)
+	case model.DatabaseDriverPostgres:
+		driver, err = ps.WithInstance(s.wDB)
+	default:
+		err = fmt.Errorf("unsupported database type %s for migration", s.driverName)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to get driver for migration: %w", err)
+	}
+
+	opts := []morph.EngineOption{
+		morph.WithLogger(log.New(&morphLogWriter{
+			log: s.log,
+		}, "", log.Lshortfile)),
+		morph.WithLock("mm-calls-migrations-lock-key"),
+		morph.SetMigrationTableName(migrationsTableName),
+		morph.SetStatementTimeoutInSeconds(timeoutSecs),
+		morph.SetDryRun(dryRun),
+	}
+
+	engine, err := morph.New(context.Background(), driver, src, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to creare morph engine: %w", err)
+	}
+
+	return engine, nil
+}
+
+func (s *Store) Migrate(direction MigrationDirection, dryRun bool) error {
+	engine, err := s.initMorph(dryRun, *s.settings.MigrationsStatementTimeoutSeconds)
+	if err != nil {
+		return fmt.Errorf("failed to initialize morph: %w", err)
+	}
+	defer engine.Close()
+
+	switch direction {
+	case MigrationsDirectionDown:
```

Comment:
@isacikgoz Other than for testing purposes I am not sure what the immediate use of down migrations could be for plugins such as Calls as there isn't really an easy way to trigger them like in the case of the server (through CLI). 

Also a plugin could technically be downgraded independently to the underlying Mattermost so other than trying to make schema changes backwards compatible I am not sure how many options we have. 

Either way, I'd love to hear your thoughts on this.

Reply:
Sounds good, thanks for clarifying on this.

Reply:
~~This may be a stupid idea, considering I don't know much about morph, but since plugins can be downgraded manually pretty easily and quickly back and forth, would it be feasible to store the down migration steps in the db? Then a downgraded plugin could read the db version, and if it sees that the db's version it is above what the current plugin supports, it would read the downgrading code from the db (left by the higher level plugin version), and run it.~~ nvm, I just read Ibrahim's proposal. 😅